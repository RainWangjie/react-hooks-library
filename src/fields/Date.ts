import moment, { Moment } from 'moment'

import { IDateField, IFieldExcludeKind } from './types'
import { Value } from './Value'

export class Date extends Value<string> {
  public moment: Moment

  public get presentation() {
    return this.format()
  }

  constructor(value: string, field: IFieldExcludeKind<IDateField>) {
    super(value, field)

    this.moment = moment(value)
  }

  public format = (formatString = 'YYYY-MM-DD') => {
    return this.moment.format(formatString)
  }
}
