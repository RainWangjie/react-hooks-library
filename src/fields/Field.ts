import { IField, IFieldExcludeKind } from './types'

export class Field implements IFieldExcludeKind<IField> {
  public displayName: string
  // 显示名称

  constructor(field: IFieldExcludeKind<IField>) {
    this.displayName = field.displayName
  }
}
