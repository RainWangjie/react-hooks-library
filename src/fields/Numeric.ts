import numeral from 'numeral'

import { INumericField, IFieldExcludeKind } from './types'
import { Value } from './Value'

export class Numeric extends Value<number> {
  public numeral: Numeral
  public decimals: number
  public isPercent?: boolean

  protected get defaultFormatString(): string {
    if (this.decimals > 0) {
      return `0,0.${'0'.repeat(this.decimals)}`
    }

    return '0,0'
  }

  public get presentation() {
    return this.format()
  }

  public get value(): number {
    return this.numeral.value()
  }

  public set value(newValue: number) {
    this.numeral = numeral(newValue)
  }

  constructor(value: number, numericField: IFieldExcludeKind<INumericField>) {
    super(value, numericField)

    this.numeral = numeral(value)
    this.decimals = numericField.decimals
    this.isPercent = numericField.isPercent
  }

  public calcTrend = (compare: number) => {
    if (!compare || compare === 0) {
      return undefined
    }

    return numeral(this.value)
      .subtract(compare)
      .divide(compare)
      .value()
  }

  public format = (formatString?: string) => {
    return this.numeral.format(formatString || this.defaultFormatString)
  }
}
