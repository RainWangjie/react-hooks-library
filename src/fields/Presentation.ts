import { IFieldExcludeKind, IPresentationField } from './types'
import { Value } from './Value'

export class Presentation extends Value<string> {
  public valueMapper?: object

  get valueDisplayName(): string {
    if (this.valueMapper !== undefined) {
      return this.valueMapper[this.value]
    }

    return this.value
  }

  constructor(
    value: string,
    presentationField: IFieldExcludeKind<IPresentationField>
  ) {
    super(value, presentationField)

    this.valueMapper = presentationField.valueMapper
  }
}
