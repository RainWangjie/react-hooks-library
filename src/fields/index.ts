export * from './Numeric'
export * from './Currency'
export * from './Date'
export * from './Field'
export * from './Text'
export * from './Value'
export * from './decorateModel'
export * from './types'
