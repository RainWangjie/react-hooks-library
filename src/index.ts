export * from './fields'
export * from './hooks'
