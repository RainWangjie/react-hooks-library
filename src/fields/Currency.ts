import numeral from 'numeral'

import { Numeric } from './Numeric'

export class Currency extends Numeric {
  public get presentation() {
    return this.format()
  }

  public get yuan(): number {
    return numeral(this.numeral)
      .divide(100)
      .value()
  }

  public get kind() {
    return Currency
  }

  public static isCurrency<P>(model: P) {
    return model instanceof Currency
  }

  public format = (formatString?: string) => {
    return numeral(this.yuan).format(formatString || this.defaultFormatString)
  }

  public clone = (value?: number) => {
    return new Currency(value || this.value, {
      decimals: this.decimals,
      displayName: this.field.displayName,
      isPercent: this.isPercent,
    })
  }
}
