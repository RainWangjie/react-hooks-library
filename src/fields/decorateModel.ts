import { isObject, isArray, isNumber, isString, isRegExp } from 'lodash'

import {
  INumericField,
  ICurrencyField,
  IPresentationField,
  IDateField,
  FieldKind,
} from './types'
import { Date } from './Date'
import { Currency } from './Currency'
import { Numeric } from './Numeric'
import { Text } from './Text'

export interface IDefintion {
  [key: string]:
    | INumericField
    | ICurrencyField
    | IPresentationField
    | IDateField
}

export function decorateModel(
  data: object | object[],
  defintions: IDefintion
): object | object[] {
  if (isObject(data)) {
    Object.keys(data).forEach(key => {
      const value = data[key]
      const defintion = defintions[key]

      if (isNumber(value) || isString(value) || isRegExp(value)) {
        if (defintion) {
          data[key] = createModel(value, defintion)
        }
      } else {
        data[key] = decorateModel(value, defintions)
      }
    })

    return data
  } else if (isArray(data)) {
    return (data as object[]).map(item => decorateModel(item, defintions))
  }

  return data
}

function createModel(value: any, defintion: IDefintion[keyof IDefintion]): any {
  switch (defintion.kind) {
    case FieldKind.Date:
      return new Date(value, {
        displayName: defintion.displayName,
      })
    case FieldKind.Currency:
      return new Currency(value, {
        displayName: defintion.displayName,
        decimals: defintion.decimals,
        isPercent: defintion.isPercent,
      })
    case FieldKind.Numeric:
      return new Numeric(value, {
        displayName: defintion.displayName,
        decimals: defintion.decimals,
        isPercent: defintion.isPercent,
      })
    case FieldKind.Text:
      return new Text(value, {
        displayName: defintion.displayName,
        valueMapper: defintion.valueMapper,
      })
    default:
      return value
  }
}
