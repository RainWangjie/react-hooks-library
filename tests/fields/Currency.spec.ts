import { Currency } from '../../src/fields'

describe('Model Currency', () => {
  const currency = new Currency(100, {
    decimals: 2,
    isPercent: false,
    displayName: '测试',
  })

  it('should create instance', () => {
    expect(JSON.stringify(currency)).toMatchSnapshot()
  })

  it('should have yuan', () => {
    expect(currency.yuan).toBe(1)
  })

  it('should have format function', () => {
    expect(currency.format()).toBe('1.00')

    currency.value = 1000000

    expect(currency.format()).toBe('10,000.00')
    expect(currency.format('0.0')).toBe('10000.0')
  })
})
