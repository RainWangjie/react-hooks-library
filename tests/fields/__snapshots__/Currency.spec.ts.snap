// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model Currency should create instance 1`] = `"{\\"numeral\\":{\\"_input\\":100,\\"_value\\":100},\\"field\\":{\\"displayName\\":\\"测试\\"},\\"decimals\\":2,\\"isPercent\\":false}"`;
